{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"POLARS_MAX_THREADS\"] = \"128\"\n",
    "\n",
    "import polars as pl\n",
    "import polars.selectors as cs\n",
    "\n",
    "pl.enable_string_cache()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python implementation: CPython\n",
      "Python version       : 3.11.3\n",
      "IPython version      : 8.13.2\n",
      "\n",
      "polars: 0.20.6\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -vp polars"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Protein-protein alignments are computed using `mmseqs2` (v13.45111):\n",
    "\n",
    "```bash\n",
    "mmseqs createdb FASTAFILE seqDB\n",
    "mmseqs search seqDB seqDB resDB tmp -s 7.5 -c 0.3 # be sure to set your thread count!\n",
    "mmseqs convertalis seqDB seqDB resDB RESULTSFILE --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qlen,tlen,qcov,tcov \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = {\n",
    "    \"query\": pl.String,\n",
    "    \"target\": pl.String,\n",
    "    \"fident\": pl.Float32,\n",
    "    \"alnlen\": pl.Int16,\n",
    "    \"mismatch\": pl.Int16,\n",
    "    \"gapopen\": pl.Int16,\n",
    "    \"qstart\": pl.Int16,\n",
    "    \"qend\": pl.Int16,\n",
    "    \"tstart\": pl.Int16,\n",
    "    \"tend\": pl.Int16,\n",
    "    \"evalue\": pl.Float32,\n",
    "    \"bits\": pl.Float32,\n",
    "    \"qlen\": pl.Int16,\n",
    "    \"tlen\": pl.Int16,\n",
    "    \"qcov\": pl.Float32,\n",
    "    \"tcov\": pl.Float32,\n",
    "}\n",
    "\n",
    "def get_genome(col: str, is_train_dataset: bool = False) -> pl.Expr:\n",
    "    # basically all proteins are named like this: scaffold_1...scaffold_2...\n",
    "    # however, for multi-scaffold viruses (vMAGs), \n",
    "    # just take the first IMGVR4 identifier to get all scaffolds\n",
    "    if is_train_dataset:\n",
    "        return pl.col(col).str.extract(r\"(.*)?_\\d+$\")\n",
    "    \n",
    "    return pl.col(col).str.split(\"|\").list.get(0).cast(pl.Categorical)\n",
    "\n",
    "def read_protein_alignments(file: str, test_run: bool=True) -> pl.DataFrame:\n",
    "    \"\"\"\n",
    "    File is a tab-separated file with the following columns:\n",
    "    query, target, fident, alnlen, mismatch, gapopen, qstart, qend, tstart, tend, evalue, bits, qlen, tlen, qcov, tcov\n",
    "\n",
    "    This is generated by mmseqs2 (v13.45111):\n",
    "        ```bash\n",
    "        mmseqs convertalis seqDB seqDB resDB results.tsv \\\n",
    "               --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qlen,tlen,qcov,tcov\n",
    "        ```\n",
    "    \"\"\"\n",
    "    data = (\n",
    "        pl.scan_csv(\n",
    "            file,\n",
    "            has_header=False,\n",
    "            separator=\"\\t\",\n",
    "            new_columns=list(cols.keys()),\n",
    "            dtypes=cols,\n",
    "            n_rows=5_000_000 if test_run else None,\n",
    "        )\n",
    "        .drop([\"mismatch\", \"gapopen\"])\n",
    "        .with_columns(\n",
    "            query_genome=get_genome(\"query\"),\n",
    "            target_genome=get_genome(\"target\"),\n",
    "        )\n",
    "        .filter(pl.col(\"query_genome\") != pl.col(\"target_genome\"))\n",
    "        .collect()\n",
    "        .cast({\"query\": pl.Categorical, \"target\": pl.Categorical})\n",
    "    )\n",
    "\n",
    "    return data\n",
    "\n",
    "def read_genome_info(file: str) -> pl.DataFrame:\n",
    "    \"\"\"\n",
    "    File is a simple tab-delimited file with 2 columns:\n",
    "        - genome: genome name\n",
    "        - ptns: number of encoded proteins\n",
    "\n",
    "    Notably, the genome name is the same as defined above in the \n",
    "    `get_genome` function.\n",
    "    \"\"\"\n",
    "    genome_info = (\n",
    "        pl.read_csv(\n",
    "            file, \n",
    "            separator=\"\\t\", \n",
    "            columns=[\"genome\", \"num_proteins\"]\n",
    "        )\n",
    "        .cast({\"genome\": pl.Categorical, \"num_proteins\": pl.Int16})\n",
    "    )\n",
    "\n",
    "    return genome_info\n",
    "\n",
    "def compute_aai(ptn_aln_file: str, genome_info_file: str, test_run: bool = False) -> pl.DataFrame:\n",
    "    ptn_alns = read_protein_alignments(ptn_aln_file, test_run)\n",
    "    genome_info = read_genome_info(genome_info_file)\n",
    "\n",
    "    aai = (\n",
    "        ptn_alns\n",
    "        .lazy()\n",
    "        # filter genome-genome comps to choose best match \n",
    "        # for each ptn from each genome. This basically only evaluates AAI\n",
    "        # for the best hits between each pair of genomes\n",
    "        .with_columns(\n",
    "            query_best_bits = pl.col(\"bits\").max().over(\"query\", \"query_genome\", \"target_genome\"),\n",
    "            query_best_cov = pl.col(\"qcov\").max().over(\"query\", \"query_genome\", \"target_genome\"),\n",
    "            target_best_bits = pl.col(\"bits\").max().over(\"target\", \"query_genome\", \"target_genome\"),\n",
    "            target_best_cov = pl.col(\"tcov\").max().over(\"target\", \"query_genome\", \"target_genome\"),\n",
    "        )\n",
    "        .filter(\n",
    "            (pl.col(\"bits\") == pl.col(\"query_best_bits\")) &\n",
    "            (pl.col(\"bits\") == pl.col(\"target_best_bits\")) &\n",
    "            (pl.col(\"qcov\") == pl.col(\"query_best_cov\")) &\n",
    "            (pl.col(\"tcov\") == pl.col(\"target_best_cov\"))\n",
    "        )\n",
    "\n",
    "        ### then compute aai\n",
    "        .group_by([\"query_genome\", \"target_genome\"])\n",
    "        .agg(\n",
    "            aai = pl.col(\"fident\").mean(),\n",
    "            shared_genes = pl.min_horizontal(pl.n_unique(\"query\", \"target\")),\n",
    "        )\n",
    "        .collect()\n",
    "        # then we need to compute the proportion of shared genes used to compute AAI\n",
    "        .join(genome_info, left_on=\"query_genome\", right_on=\"genome\")\n",
    "        .rename({\"num_proteins\": \"query_ptns\"})\n",
    "        .join(genome_info, left_on=\"target_genome\", right_on=\"genome\")\n",
    "        .rename({\"num_proteins\": \"target_ptns\"})\n",
    "        .with_columns(\n",
    "            query_shared = (pl.col(\"shared_genes\") / pl.col(\"query_ptns\")).cast(pl.Float32),\n",
    "            target_shared = (pl.col(\"shared_genes\") / pl.col(\"target_ptns\")).cast(pl.Float32),\n",
    "        )\n",
    "    )\n",
    "\n",
    "    return aai\n",
    "\n",
    "def replace_ext(file: str, ext: str) -> str:\n",
    "    return f'{file.rsplit(\".\", 1)[0]}.{ext}'\n",
    "\n",
    "def save_aai(aai: pl.DataFrame, out_file: str, parquet: bool = True):\n",
    "    if parquet:\n",
    "        # output will generally be very large, \n",
    "        # so parquet is much more storage efficient\n",
    "        out_file = replace_ext(out_file, \"parquet\")\n",
    "        aai.write_parquet(out_file)\n",
    "    else:\n",
    "        out_file = replace_ext(out_file, \"tsv\")\n",
    "        aai.write_csv(out_file, separator=\"\\t\")\n",
    "\n",
    "    print(f\"Saved AAI to {out_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (4_574_151, 8)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>query_genome</th><th>target_genome</th><th>aai</th><th>shared_genes</th><th>query_ptns</th><th>target_ptns</th><th>query_shared</th><th>target_shared</th></tr><tr><td>cat</td><td>cat</td><td>f32</td><td>u32</td><td>i16</td><td>i16</td><td>f32</td><td>f32</td></tr></thead><tbody><tr><td>&quot;IMGVR_UViG_271…</td><td>&quot;IMGVR_UViG_649…</td><td>0.873</td><td>1</td><td>121</td><td>133</td><td>0.008264</td><td>0.007519</td></tr><tr><td>&quot;IMGVR_UViG_271…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>121</td><td>90</td><td>0.008264</td><td>0.011111</td></tr><tr><td>&quot;IMGVR_UViG_271…</td><td>&quot;IMGVR_UViG_330…</td><td>0.329</td><td>1</td><td>121</td><td>76</td><td>0.008264</td><td>0.013158</td></tr><tr><td>&quot;IMGVR_UViG_280…</td><td>&quot;IMGVR_UViG_281…</td><td>0.282</td><td>1</td><td>51</td><td>70</td><td>0.019608</td><td>0.014286</td></tr><tr><td>&quot;IMGVR_UViG_280…</td><td>&quot;IMGVR_UViG_330…</td><td>0.272</td><td>1</td><td>51</td><td>71</td><td>0.019608</td><td>0.014085</td></tr><tr><td>&quot;IMGVR_UViG_280…</td><td>&quot;IMGVR_UViG_330…</td><td>0.32</td><td>1</td><td>51</td><td>90</td><td>0.019608</td><td>0.011111</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_268…</td><td>0.714</td><td>1</td><td>41</td><td>42</td><td>0.02439</td><td>0.02381</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.704</td><td>1</td><td>41</td><td>39</td><td>0.02439</td><td>0.025641</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_275…</td><td>0.632</td><td>1</td><td>41</td><td>46</td><td>0.02439</td><td>0.021739</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_289…</td><td>0.592</td><td>1</td><td>41</td><td>44</td><td>0.02439</td><td>0.022727</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_284…</td><td>0.261</td><td>1</td><td>211</td><td>82</td><td>0.004739</td><td>0.012195</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.237</td><td>1</td><td>211</td><td>78</td><td>0.004739</td><td>0.012821</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.321</td><td>1</td><td>63</td><td>71</td><td>0.015873</td><td>0.014085</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.354</td><td>1</td><td>228</td><td>393</td><td>0.004386</td><td>0.002545</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.205</td><td>1</td><td>228</td><td>61</td><td>0.004386</td><td>0.016393</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.537</td><td>1</td><td>114</td><td>143</td><td>0.008772</td><td>0.006993</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.53</td><td>1</td><td>114</td><td>132</td><td>0.008772</td><td>0.007576</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.572</td><td>1</td><td>114</td><td>126</td><td>0.008772</td><td>0.007937</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.482</td><td>1</td><td>114</td><td>59</td><td>0.008772</td><td>0.016949</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.415</td><td>1</td><td>114</td><td>65</td><td>0.008772</td><td>0.015385</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.342</td><td>1</td><td>114</td><td>138</td><td>0.008772</td><td>0.007246</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_288…</td><td>0.447</td><td>1</td><td>54</td><td>59</td><td>0.018519</td><td>0.016949</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.645</td><td>1</td><td>60</td><td>65</td><td>0.016667</td><td>0.015385</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.417</td><td>1</td><td>60</td><td>53</td><td>0.016667</td><td>0.018868</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (4_574_151, 8)\n",
       "┌────────────┬────────────┬───────┬────────────┬────────────┬────────────┬────────────┬────────────┐\n",
       "│ query_geno ┆ target_gen ┆ aai   ┆ shared_gen ┆ query_ptns ┆ target_ptn ┆ query_shar ┆ target_sha │\n",
       "│ me         ┆ ome        ┆ ---   ┆ es         ┆ ---        ┆ s          ┆ ed         ┆ red        │\n",
       "│ ---        ┆ ---        ┆ f32   ┆ ---        ┆ i16        ┆ ---        ┆ ---        ┆ ---        │\n",
       "│ cat        ┆ cat        ┆       ┆ u32        ┆            ┆ i16        ┆ f32        ┆ f32        │\n",
       "╞════════════╪════════════╪═══════╪════════════╪════════════╪════════════╪════════════╪════════════╡\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.873 ┆ 1          ┆ 121        ┆ 133        ┆ 0.008264   ┆ 0.007519   │\n",
       "│ _271821842 ┆ _649989997 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 0_000001   ┆ _000001    ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.4   ┆ 1          ┆ 121        ┆ 90         ┆ 0.008264   ┆ 0.011111   │\n",
       "│ _271821842 ┆ _330002427 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 0_000001   ┆ 0_003055   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.329 ┆ 1          ┆ 121        ┆ 76         ┆ 0.008264   ┆ 0.013158   │\n",
       "│ _271821842 ┆ _330002654 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 0_000001   ┆ 9_001205   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.282 ┆ 1          ┆ 51         ┆ 70         ┆ 0.019608   ┆ 0.014286   │\n",
       "│ _280242882 ┆ _281633228 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 6_000002   ┆ 9_000001   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.272 ┆ 1          ┆ 51         ┆ 71         ┆ 0.019608   ┆ 0.014085   │\n",
       "│ _280242882 ┆ _330004598 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 6_000002   ┆ 8_024903   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ …          ┆ …          ┆ …     ┆ …          ┆ …          ┆ …          ┆ …          ┆ …          │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.415 ┆ 1          ┆ 114        ┆ 65         ┆ 0.008772   ┆ 0.015385   │\n",
       "│ _330000915 ┆ _330003569 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 8_000037   ┆ 8_004282   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.342 ┆ 1          ┆ 114        ┆ 138        ┆ 0.008772   ┆ 0.007246   │\n",
       "│ _330000915 ┆ _330001648 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 8_000037   ┆ 4_000447   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.447 ┆ 1          ┆ 54         ┆ 59         ┆ 0.018519   ┆ 0.016949   │\n",
       "│ _330001266 ┆ _288211479 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 4_000397   ┆ 9_000001   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.645 ┆ 1          ┆ 60         ┆ 65         ┆ 0.016667   ┆ 0.015385   │\n",
       "│ _330001266 ┆ _330002542 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 4_000189   ┆ 6_000070   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.417 ┆ 1          ┆ 60         ┆ 53         ┆ 0.016667   ┆ 0.018868   │\n",
       "│ _330001266 ┆ _330004340 ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "│ 4_000189   ┆ 9_000243   ┆       ┆            ┆            ┆            ┆            ┆            │\n",
       "└────────────┴────────────┴───────┴────────────┴────────────┴────────────┴────────────┴────────────┘"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_info_file = \"supplementary_tables/supplementary_table_1.tsv\"\n",
    "\n",
    "# this file is provided gzipped because it is extremely large (>900M rows, 170GB)\n",
    "# so you will need to unzip it before running this code\n",
    "# but you can generate this file with the above mmseqs commands and the\n",
    "# test set protein FASTA file that is provided\n",
    "ptn_aln_file = \"datasets/aai/test_set_protein-protein_alignments.tsv\"\n",
    "\n",
    "aai = compute_aai(ptn_aln_file, genome_info_file, test_run=True)\n",
    "aai"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Virus clustering based on AAI\n",
    "\n",
    "We can prune the above pairwise AAI calculations to only consider clustering viruses at broader evolutionary connections. The following thresholds were used [previously](https://github.com/snayfach/MGV/tree/master/aai_cluster), but you could use your own thresholds.\n",
    "\n",
    "We view this table as a graph, where the nodes are the genomes, and edges connect genomes based on detectable AAI. The edge weight is just the AAI and minimum proportion of shared genes. This notably penalizes high AAI based on only a few shared genes. Thus, any downstream clusters will be viruses that are more holistically related."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_for_clustering(aai_df: pl.DataFrame, cluster_level: str) -> pl.DataFrame:\n",
    "    if cluster_level == \"genus\":\n",
    "        min_aai = 0.4\n",
    "        min_shared = 0.2\n",
    "        min_genes = 16\n",
    "    elif cluster_level == \"family\":\n",
    "        min_aai = 0.2\n",
    "        min_shared = 0.1\n",
    "        min_genes = 8\n",
    "    else:\n",
    "        raise ValueError(f\"Invalid cluster level: {cluster_level} | Must be 'genus' or 'family'\")\n",
    "    \n",
    "    filtered_aai = (\n",
    "        aai_df\n",
    "        .lazy()\n",
    "        .with_columns(\n",
    "            min_shared = pl.min_horizontal(cs.ends_with(\"_shared\")),\n",
    "        )\n",
    "        .filter(\n",
    "            (pl.col(\"aai\") >= min_aai) &\n",
    "            (\n",
    "                (pl.col(\"min_shared\") >= min_shared) |\n",
    "                (pl.col(\"shared_genes\") >= min_genes)\n",
    "            )\n",
    "        )\n",
    "        .with_columns(\n",
    "            edge_weight = pl.col(\"aai\") * pl.col(\"min_shared\"),\n",
    "        )\n",
    "        .sort(\"edge_weight\", descending=True)\n",
    "        .collect()\n",
    "    )\n",
    "\n",
    "    return filtered_aai\n",
    "\n",
    "def save_for_clustering(filtered_aai: pl.DataFrame, out_file: str):\n",
    "    out_file = replace_ext(out_file, \"abc\")\n",
    "    print(f\"Saving for mcl clustering to {out_file}\")\n",
    "\n",
    "    (\n",
    "        filtered_aai\n",
    "        .select(\"query_genome\", \"target_genome\", \"edge_weight\")\n",
    "        .write_csv(out_file, separator=\"\\t\", include_header=False)\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The only columns needed for clustering (which define a text-based tabular graph) are `query_genome`, `target_genome`, and `edge_weight`. The above function `save_for_clustering` will save the dataframe for you.\n",
    "\n",
    "Note: Throughout the manuscript, we refer to \"scaled AAI\", which is the `edge_weight` here. This is AAI scaled by the minimum proportion of shared genes used to compute AAI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (42_080, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>query_genome</th><th>target_genome</th><th>aai</th><th>shared_genes</th><th>query_ptns</th><th>target_ptns</th><th>query_shared</th><th>target_shared</th><th>min_shared</th><th>edge_weight</th></tr><tr><td>cat</td><td>cat</td><td>f32</td><td>u32</td><td>i16</td><td>i16</td><td>f32</td><td>f32</td><td>f32</td><td>f32</td></tr></thead><tbody><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>1.0</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.5</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.976</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.488</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.967</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4835</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.959</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4795</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.956</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.478</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.953</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4765</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.945</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4725</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.937</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4685</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.935</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4675</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.919</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.4595</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.904</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.452</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.902</td><td>1</td><td>2</td><td>2</td><td>0.5</td><td>0.5</td><td>0.5</td><td>0.451</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>2</td><td>5</td><td>0.5</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>4</td><td>5</td><td>0.25</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>4</td><td>0.2</td><td>0.25</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>4</td><td>5</td><td>0.25</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>3</td><td>0.2</td><td>0.333333</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>5</td><td>0.2</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>4</td><td>0.2</td><td>0.25</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>5</td><td>0.2</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>4</td><td>5</td><td>0.25</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>2</td><td>0.2</td><td>0.5</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>5</td><td>5</td><td>0.2</td><td>0.2</td><td>0.2</td><td>0.08</td></tr><tr><td>&quot;IMGVR_UViG_330…</td><td>&quot;IMGVR_UViG_330…</td><td>0.4</td><td>1</td><td>4</td><td>5</td><td>0.25</td><td>0.2</td><td>0.2</td><td>0.08</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (42_080, 10)\n",
       "┌────────────┬────────────┬───────┬────────────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ query_geno ┆ target_gen ┆ aai   ┆ shared_gen ┆ … ┆ query_sha ┆ target_sh ┆ min_share ┆ edge_weig │\n",
       "│ me         ┆ ome        ┆ ---   ┆ es         ┆   ┆ red       ┆ ared      ┆ d         ┆ ht        │\n",
       "│ ---        ┆ ---        ┆ f32   ┆ ---        ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │\n",
       "│ cat        ┆ cat        ┆       ┆ u32        ┆   ┆ f32       ┆ f32       ┆ f32       ┆ f32       │\n",
       "╞════════════╪════════════╪═══════╪════════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 1.0   ┆ 1          ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.5       │\n",
       "│ _330002004 ┆ _330002001 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 7_015909   ┆ 7_000804   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.976 ┆ 1          ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.488     │\n",
       "│ _330001014 ┆ _330003611 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 4_000642   ┆ 1_000001   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.967 ┆ 1          ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.4835    │\n",
       "│ _330002252 ┆ _330002271 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 2_000131   ┆ 8_000149   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.959 ┆ 1          ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.4795    │\n",
       "│ _330001277 ┆ _330001278 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 8_000030   ┆ 1_000104   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.956 ┆ 1          ┆ … ┆ 0.5       ┆ 0.5       ┆ 0.5       ┆ 0.478     │\n",
       "│ _330000623 ┆ _330000622 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 1_000281   ┆ 8_000198   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ …          ┆ …          ┆ …     ┆ …          ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.4   ┆ 1          ┆ … ┆ 0.2       ┆ 0.2       ┆ 0.2       ┆ 0.08      │\n",
       "│ _330003466 ┆ _330003109 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 1_000142   ┆ 9_000133   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.4   ┆ 1          ┆ … ┆ 0.25      ┆ 0.2       ┆ 0.2       ┆ 0.08      │\n",
       "│ _330003875 ┆ _330002002 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 8_000110   ┆ 8_000594   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.4   ┆ 1          ┆ … ┆ 0.2       ┆ 0.5       ┆ 0.2       ┆ 0.08      │\n",
       "│ _330003466 ┆ _330004741 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 6_000220   ┆ 8_000097   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.4   ┆ 1          ┆ … ┆ 0.2       ┆ 0.2       ┆ 0.2       ┆ 0.08      │\n",
       "│ _330001035 ┆ _330002007 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 6_002889   ┆ 8_001189   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ IMGVR_UViG ┆ IMGVR_UViG ┆ 0.4   ┆ 1          ┆ … ┆ 0.25      ┆ 0.2       ┆ 0.2       ┆ 0.08      │\n",
       "│ _330000373 ┆ _330003466 ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "│ 5_000078   ┆ 6_000220   ┆       ┆            ┆   ┆           ┆           ┆           ┆           │\n",
       "└────────────┴────────────┴───────┴────────────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_aai = prepare_for_clustering(aai, \"genus\")\n",
    "filtered_aai"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can use any graph clustering method you prefer, but for the PST manuscript, we clustered genomes on the basis of AAI using the Markov clustering algorithm ([mcl](https://github.com/micans/mcl)).\n",
    "\n",
    "We use the following setup for mcl clustering:\n",
    "\n",
    "```bash\n",
    "mcxload -abc AAI_GRAPH.abc -o AAI_GRAPH.mci -write-tab AAI_GRAPH.tab\n",
    "mcl AAI_GRAPH.mci -I 2.0 -use-tab AAI_GRAPH.tab -o AAI_genome_clusters.txt\n",
    "```\n",
    "\n",
    "We toggled the mcl inflation value (`-I`) to be 2.0 for \"medium\" resolution and 4.0 for \"high\" resolution.\n",
    "\n",
    "Where `AAI_GRAPH.abc` is the file produced from the `save_for_clustering` function."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datasci",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
